src/interop.ts

```diff
@@ -12,25 +12,304 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import * as cdk from 'aws-cdk-lib';
 import * as pulumi from '@pulumi/pulumi';
 import { debug } from '@pulumi/pulumi/log';
 import { IConstruct } from 'constructs';
-import { moduleName, toSdkName, typeToken } from './naming';
+import { toSdkName, typeToken } from './naming';
+import * as path from 'path';
+import * as fs from 'fs';
+
+const glob = global as any;
 
 export function firstToLower(str: string) {
     return str.replace(/\w\S*/g, function (txt) {
-        return txt.charAt(0).toLowerCase() + txt.substr(1);
+        return txt.charAt(0).toLowerCase() + txt.substring(1);
     });
 }
 
-export function normalize(value: any): any {
+class UnknownCfnType extends Error {
+    constructor() {
+        super("CfnType doesn't exist as a native type");
+    }
+}
+
+export class Metadata {
+    public static instance(provider: PulumiProvider): Metadata {
+        if (glob.__pulumiMetadata == undefined) {
+            glob.__pulumiMetadata = new Metadata(provider);
+        }
+        return glob.__pulumiMetadata;
+    }
+    private readonly pulumiMetadata: PulumiMetadata;
+    constructor(provider: PulumiProvider) {
+        if (provider === PulumiProvider.AWS_NATIVE) {
+            this.pulumiMetadata = this.readMetadata();
+        } else {
+            throw new Error('AWS_NATIVE is the only supported pulumi provider');
+        }
+    }
+
+    /**
+     * Finds a specific resource in the metadata. Throws an UnknownCfnType error
+     * if one is not found (probably means it is a classic type).
+     *
+     * @param cfnType The CloudFormation type i.e. AWS::S3::Bucket
+     * @param metadata The PulumiMetadata to search
+     * @returns The PulumiResource
+     * @throws UnknownCfnType if the resource is not found
+     */
+    public findResource(cfnType: string): PulumiResource {
+        const pType = typeToken(cfnType);
+        if (pType in this.pulumiMetadata.resources) {
+            return this.pulumiMetadata.resources[pType];
+        }
+        throw new UnknownCfnType();
+    }
+
+    public types(): { [key: string]: PulumiType } {
+        return this.pulumiMetadata.types;
+    }
+
+    /**
+     * Read the aws-native metadata.json file
+     *
+     * @returns the PulumiMetadata
+     */
+    private readMetadata(): PulumiMetadata {
+        const contents = fs.readFileSync(path.join(__dirname, '../schemas/aws-native-metadata.json'), {
```

Comment:
Will this file be available every place this code needs to ship?

Reply:
Currently this library is node only and this file is shipped as part of the npm package.

Reply:
Nice!